aws/resource_aws_opsworks_chef.go

```diff
@@ -0,0 +1,335 @@
+		PreferredMaintenanceWindow: aws.String(d.Get("preferred_maintenance_window").(string)),
+		SecurityGroupIds:           aws.StringSlice(d.Get("security_group_ids").([]string)), // TODO: SecurityGroupIds, set?
+		ServiceRoleArn:             aws.String(d.Get("service_role_arn").(string)),
+		SubnetIds:                  aws.StringSlice(d.Get("subnet_ids").([]string)), // TODO: set?
+	}
+
+	chefPivotalKeyAttribute := opsworkscm.EngineAttribute{
+		Name:  aws.String("CHEF_PIVOTAL_KEY"),
+		Value: aws.String(d.Get("chef_pivotal_key").(string)),
+	}
+
+	chefDeliveryAdminPassword := opsworkscm.EngineAttribute{
+		Name:  aws.String("CHEF_DELIVERY_ADMIN_PASSWORD"),
+		Value: aws.String(d.Get("chef_delivery_admin_password").(string)),
+	}
+
+	req.SetEngineAttributes([]*opsworkscm.EngineAttribute{&chefPivotalKeyAttribute, &chefDeliveryAdminPassword})
+
+	// TODO: possibly make these optional
+	// TODO: security_group_ids, optional
+	// TODO: subnet_ids, optional
+
+	if sshKeyPair, ok := d.GetOk("ssh_key_pair"); ok {
+		req.KeyPair = aws.String(sshKeyPair.(string))
+	}
+
+	log.Printf("[DEBUG] Creating OpsWorks Chef server: %s", req)
+
+	var resp *opsworkscm.CreateServerOutput
+	err = resource.Retry(20*time.Minute, func() *resource.RetryError {
+		var cerr error
+		resp, cerr = client.CreateServer(req)
+		if cerr != nil {
+			if opserr, ok := cerr.(awserr.Error); ok {
+				// TODO: does this also happen with this resource?
+				// If Terraform is also managing the service IAM role,
+				// it may have just been created and not yet be
+				// propagated.
+				// AWS doesn't provide a machine-readable code for this
+				// specific error, so we're forced to do fragile message
+				// matching.
+				// The full error we're looking for looks something like
+				// the following:
+				// Service Role Arn: [...] is not yet propagated, please try again in a couple of minutes
+				propErr := "not yet propagated"
+				trustErr := "not the necessary trust relationship"
+				validateErr := "validate IAM role permission"
+				if opserr.Code() == "ValidationException" && (strings.Contains(opserr.Message(), trustErr) || strings.Contains(opserr.Message(), propErr) || strings.Contains(opserr.Message(), validateErr)) {
+					log.Printf("[INFO] Waiting for service IAM role to propagate")
+					return resource.RetryableError(cerr)
+				}
+			}
+			return resource.NonRetryableError(cerr)
+		}
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+
+	serverName := *resp.Server.ServerName
+	d.SetId(serverName)
+
+	return resourceAwsOpsworksChefUpdate(d, meta)
+}
+
+func resourceAwsOpsworksChefUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*AWSClient).opsworkscmconn
+
+	err := resourceAwsOpsworksChefValidate(d)
+	if err != nil {
+		return err
+	}
+
+	// TODO: these are the only values that are allowed to be changed
+	// according to the API, is there a way to express this in the provider somehow?
```

Comment:
question here: only some of the values are able to be updated. The others are write-once. How to handle?

There are actually 2 more that can be updated: the `chef_pivotal_key` and `chef_delivery_admin_password`, but those are actually set using a different API call (easy enough to handle, though) but the question there is how do we know if those values need to change. Is there a way we can store the output of the createserver in a hashed form somehow so we can compare the inputs to what we last knew the values were on the server and update them? Or something else?

Reply:
Seems like `ForceNew` on the attribute is the Schema is what I'm after for the first part of that question.

Reply:
followup question on the ForceNew: will that propagate to any node associations that exist? Also, I should probably create a resource for those.

https://docs.aws.amazon.com/sdk-for-go/api/service/opsworkscm/#OpsWorksCM.AssociateNode

Reply:
hmm. Now that I think of it, the AssociateNode is probably going to be done outside of terraform scope, which means terraform wouldn't have the info needed to be able to recreate the associations. So a ForceNew would be a pretty destructive thing.

I'm curious how this is even handled on AWS side. If instance type is one of the forcenew params, then I'm assuming they're not able to roll the backend chef machine(s) to new sizes, which makes me wonder if they even have multiples or if it's HA or whatever, what happens when one dies? I'm probably reading too much into that part of things, but I might contact amazon support and ask for more details.

Reply:
Ok, I filed a support request with amazon to see what they have to say about this. For now, I'll leave ForceNew on, but I actually think forcing new here would be a really bad idea, so I'm hoping there are better ways to handle that situation :) I'm open to any suggestions.